book-content/chapters/09-typescript-only-features.md

```diff
@@ -0,0 +1,518 @@
+
+```ts twoslash
+// @errors: 2345
+enum AlbumStatus {
+  NewRelease = 0,
+  OnSale = 1,
+  StaffPick = 2,
+}
+
+function logStatus(genre: AlbumStatus) {
+  console.log(genre);
+}
+
+// ---cut---
+logStatus(3);
+```
+
+This is different from string enums, which only allow the enum members to be used as types:
+
+```ts twoslash
+// @errors: 2345
+enum AlbumStatus {
+  NewRelease = "NEW_RELEASE",
+  OnSale = "ON_SALE",
+  StaffPick = "STAFF_PICK",
+}
+
+function logStatus(genre: AlbumStatus) {
+  console.log(genre);
+}
+
+logStatus(AlbumStatus.NewRelease);
+logStatus("NEW_RELEASE");
+```
+
+The way string enums behave feels more natural - it matches how enums work in other languages like C# and Java.
+
+But the fact that they're not consistent with numeric enums can be a source of confusion.
+
+In fact, string enums are unique in TypeScript because they're compared _nominally_. All other types in TypeScript are compared _structurally_, meaning that two types are considered the same if they have the same structure. But string enums are compared based on their name (nominally), not their structure.
+
+This means that two string enums with the same members are considered different types if they have different names:
+
+```ts twoslash
+// @errors: 2345
+enum AlbumStatus {
+  NewRelease = "NEW_RELEASE",
+  OnSale = "ON_SALE",
+  StaffPick = "STAFF_PICK",
+}
+function logStatus(genre: AlbumStatus) {
+  console.log(genre);
+}
+
+// ---cut---
+enum AlbumStatus2 {
+  NewRelease = "NEW_RELEASE",
+  OnSale = "ON_SALE",
+  StaffPick = "STAFF_PICK",
+}
+
+logStatus(AlbumStatus2.NewRelease);
+```
+
+For those of us used to structural typing, this can be a bit of a surprise. But to developers used to enums in other languages, string enums will feel the most natural.
+
+#### `const` Enums
+
+A `const` enum is declared similarly to the other enums, but with the `const` keyword first:
+
+```typescript
+const enum AlbumStatus {
+  NewRelease = "NEW_RELEASE",
+  OnSale = "ON_SALE",
+  StaffPick = "STAFF_PICK",
+}
+```
+
+You can use `const` enums to declare numeric or string enums - they have the same behavior as regular enums.
+
+The major difference is that `const` enums disappear when the TypeScript is transpiled to JavaScript. Instead of creating an object with the enum's values, the transpiled JavaScript will use the enum's values directly.
+
+For instance, if an array is created that accesses the enum's values, the transpiled JavaScript will end up with those values:
+
+```typescript
+let albumStatuses = [
+  AlbumStatus.NewRelease,
+  AlbumStatus.OnSale,
+  AlbumStatus.StaffPick,
+];
+
+// the above transpiles to:
+let albumStatuses = ["NEW_RELEASE", "ON_SALE", "STAFF_PICK"];
+```
+
+`const` enums do have some limitations, especially when declared in declaration files (which we'll cover later). The TypeScript team actually recommends avoiding `const` enums in your library code because they can behave unpredictably for consumers of your library.
```

Comment:
it might be worth calling out that one of the major limitations is that only TS can emit code using them, some other tools might have partial support too (IIRC `esbuild` handles them - at least to some extent, I'm not sure) but they are not guaranteed to work if using other tools. But maybe that goes outside of the scope of this book, dunno